maro/simulator/scenarios/supply_chain/units/storage.py

```diff
@@ -1,51 +1,219 @@
+    def _take_product(self, product_id: int, quantity: int) -> None:
+        assert self._product_level[product_id] >= quantity
+        self._product_level[product_id] -= quantity
+        self._product_level_changed[product_id] = True
+        self._remaining_space_dict[self._product2storage[product_id]] += quantity
+
+    def try_add_products(
+        self,
+        product_quantities: Dict[int, int],
+        add_strategy: AddStrategy=AddStrategy.IgnoreUpperBoundAllOrNothing
+    ) -> Dict[int, int]:
+        """Try to add products into storage.
 
-            self.remaining_space -= unload_quantity
+        Args:
+            product_quantities (Dict[int, int]): Dictionary of product id and quantity need to add to storage.
+            add_strategy (AddStrategy): The strategy to add products into the storage, Defaults to AddAllOrNothing.
 
-        return unloaded_quantities
+        Returns:
+            Dict[int, int]: Dictionary of product id and quantity success added.
+        """
+        added_quantities: Dict[int, int] = {}
+
+        if add_strategy in [AddStrategy.IgnoreUpperBoundAllOrNothing, AddStrategy.IgnoreUpperBoundProportional]:
+            space_requirements: Dict[int, int] = {}
+            for product_id, quantity in product_quantities.items():
+                storage_id = self._product2storage[product_id]
+                if storage_id not in space_requirements:
+                    space_requirements[storage_id] = 0
+                space_requirements[storage_id] += quantity
+
+        if add_strategy == AddStrategy.IgnoreUpperBoundAllOrNothing:
+            for storage_id, requirement in space_requirements.items():
+                if self._remaining_space_dict[storage_id] < requirement:
+                    return {}
+
+            for product_id, quantity in product_quantities.items():
+                self._add_product(product_id, quantity)
+                added_quantities[product_id] = quantity
+
+        if add_strategy == AddStrategy.IgnoreUpperBoundProportional:
+            fulfill_ratio_dict = {}
+            for storage_id, requirement in space_requirements.items():
+                fulfill_ratio_dict[storage_id] = requirement / self._remaining_space_dict[storage_id]
+
+            for product_id, quantity in product_quantities.items():
+                storage_id = self._product2storage[product_id]
+                quantity = min(int(quantity * fulfill_ratio_dict[storage_id]), self._remaining_space_dict[storage_id])
+                self._add_product(product_id, quantity)
+                added_quantities[product_id] = quantity
+
+        if add_strategy == AddStrategy.IgnoreUpperBoundFIFO:
+            for product_id, quantity in product_quantities.items():
+                quantity = min(quantity, self._remaining_space_dict[self._product2storage[product_id]])
+                self._add_product(product_id, quantity)
+                added_quantities[product_id] = quantity
```

Comment:
Should we add `product_id` into `added_quantities` when `quantity` is zero?

Reply:
As it is the return value of this function, may be used by the caller to check one by one. I think added with 0 quantity is OK